ibis/client.py

```diff
@@ -923,6 +923,136 @@ def _get_schema_using_query(self, query):
 
         return ir.Schema(names, ibis_types)
 
+    def create_udf(self, udf_info, udf_name=None, db=None):
+        """
+        Creates a function within Impala
+
+        Parameters
+        ----------
+        udf_info : UDFInfo object
+        udf_name : string (optional)
+        db : string (optional)
+        """
+        name = udf_info.get_name()
+        if udf_name:
+            name = udf_name
+        statement = ddl.CreateFunction(udf_info.hdfs_file,
+                                       udf_info.so_symbol,
+                                       udf_info.inputs,
+                                       udf_info.output,
+                                       name, db)
+        self._execute(statement)
+
+    def create_uda(self, uda_info, uda_name=None, db=None):
+        """
+        Creates a user-defined aggregate function within Impala
+
+        Parameters
+        ----------
+        uda_info : UDAInfo object
+        uda_name : string (optional)
+        db : string (optional)
+        """
+        name = uda_info.get_name()
+        if uda_name:
+            name = uda_name
+        statement = ddl.CreateAggregateFunction(uda_info.hdfs_file,
+                                                uda_info.input_type,
+                                                uda_info.output_type,
+                                                uda_info.init_fn,
+                                                uda_info.update_fn,
+                                                uda_info.merge_fn,
+                                                uda_info.finalize_fn,
+                                                name,
+                                                db=db)
+        self._execute(statement)
+
+    def drop_udf(self, name, input_types, db=None, must_exist=False,
+                 aggregate=False):
+        """
+        Drops a function
+
+        Parameters
+        ----------
+        name : string
+        input_types : list of strings
+        must_exist : boolean, default False
+        db : string, default None
+        aggregate : boolean, default False
+        """
+        inputs = [ir._validate_type(x) for x in input_types]
+        statement = ddl.DropFunction(name, inputs, must_exist=must_exist,
+                                     aggregate=aggregate, db=db)
+        self._execute(statement)
+
+    def list_udfs(self, db, like=None):
+        """
+        Lists all UDFs associated with given database
+
+        Parameters
+        ----------
+        db : string
+        like : string for searching (optional)
+        """
+        statement = 'SHOW FUNCTIONS IN {0}'.format(db)
+        if like:
+            statement += " LIKE '{0}'".format(like)
+
+        cur = self._execute(statement)
+        result = self._get_list(cur)
+        cur.release()
+
+        return result
```

Comment:
The return type of `SHOW FUNCTIONS` is more complex than the other `SHOW` cases. What do you think about parsing the output to return a UDF metadata object – function name, database, return type, and input types. You can parse the function signature with a regular expression

http://www.cloudera.com/content/cloudera/en/documentation/cloudera-impala/latest/topics/impala_show.html#show_functions_unique_1